redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+        if self.state is no_default:
+            self.state = doc
+        # in case we need a bit more flexibility eg lambda x: np.empty(x.shape)
+        elif hasattr(self.state, '__call__'):
+            self.state = self.state(doc)
+        else:
+            doc[self.state_key] = self.state
+            result = self.func(doc)
+            self.state = result
+        return super().event(self.issue_event(self.state))
+
+
+class zip(EventStream):
+    """Combine two streams together into a stream of tuples"""
+    def __init__(self, *children, **kwargs):
+        """Initialize the Node
+
+        Parameters
+        ----------
+        children: EventStream instances
+            The event streams to be zipped together
+        """
+        self.maxsize = kwargs.pop('maxsize', 10)
+        self.buffers = [deque() for _ in children]
+        self.condition = Condition()
+        self.prior = ()
+        EventStream.__init__(self, children=children)
+
+    def update(self, x, who=None):
+        L = self.buffers[self.children.index(who)]
+        L.append(x)
+        if len(L) == 1 and all(self.buffers):
+            if self.prior:
+                for i in range(len(self.buffers)):
+                    # If the docs don't match, preempt with prior good result
+                    if self.buffers[i][0][0] != self.buffers[0][0][0]:
+                        self.buffers[i].appendleft(self.prior[i])
+            tup = tuple(buf.popleft() for buf in self.buffers)
+            self.condition.notify_all()
+            self.prior = tup
+            return self.emit(tup)
+        elif len(L) > self.maxsize:
+            return self.condition.wait()
+
+
+class bundle(EventStream):
+    """Combine multiple event streams into one"""
+    def __init__(self, *children, **kwargs):
+        """Initialize the Node
+
+        Parameters
+        ----------
+        children: EventStream instances
+            The event streams to be zipped together
+        """
+        self.maxsize = kwargs.pop('maxsize', 100)
+        self.buffers = [deque() for _ in children]
+        self.condition = Condition()
+        self.prior = ()
+        EventStream.__init__(self, children=children)
+        self.generate_provenance()
+
+    def update(self, x, who=None):
+        L = self.buffers[self.children.index(who)]
+        L.append(x)
+        if len(L) == 1 and all(self.buffers):
+            # if all the docs are of the same type and not an event, issue
+            # new documents which are combined
+            rvs = []
+            while all(self.buffers):
+                if all([b[0][0] == self.buffers[0][0][0] and b[0][0] != 'event'
+                        for b in self.buffers]):
+                    res = self.dispatch(
+                        tuple([b.popleft() for b in self.buffers]))
+                    rvs.append(self.emit(res))
+                elif any([b[0][0] == 'event' for b in self.buffers]):
```

Comment:
Essentially if there are any events next in line, we need to issue those events, hence the `elif` statement. I'll try to be a bit more explicit.